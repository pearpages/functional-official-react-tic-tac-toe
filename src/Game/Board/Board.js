import React from 'react'
import PropTypes from 'prop-types'

import './Board.css'
import { makeRenderSquareWithValueAndClickHandler } from './board.helpers'

export function Board({ config, boardRenderer }) {

  const renderSquareWithValueAndClickHandler = boardRenderer(config)

  return (
    <div data-testid="Board">
      <div className={Board.rowClassName}>
        {renderSquareWithValueAndClickHandler(0)}
        {renderSquareWithValueAndClickHandler(1)}
        {renderSquareWithValueAndClickHandler(2)}
      </div>
      <div className={Board.rowClassName}>
        {renderSquareWithValueAndClickHandler(3)}
        {renderSquareWithValueAndClickHandler(4)}
        {renderSquareWithValueAndClickHandler(5)}
      </div>
      <div className={Board.rowClassName}>
        {renderSquareWithValueAndClickHandler(6)}
        {renderSquareWithValueAndClickHandler(7)}
        {renderSquareWithValueAndClickHandler(8)}
      </div>
    </div>
  );
}

Board.rowClassName = 'board-row'

const configShape = {
  squares: PropTypes.array.isRequired,
  handleClick: PropTypes.func.isRequired,
  line: PropTypes.array
}

Board.propTypes = {
  config: PropTypes.shape(configShape).isRequired,
  boardRenderer: PropTypes.func.isRequired
}

Board.defaultProps = {
  boardRenderer: makeRenderSquareWithValueAndClickHandler
}
