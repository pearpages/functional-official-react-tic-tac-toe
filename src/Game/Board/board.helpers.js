import { Square } from './Square';

const arrayIncludesValue = (array, value) => (array || []).includes(value);

const makeArrayIncludesValue = (array) => (value) =>
  arrayIncludesValue(array, value);

export const makeRenderSquareWithValueAndClickHandler = (config) => (
  squareIndex
) => Square.renderSquare(adapter(config, squareIndex));

export const adapter = ({ squares, handleClick, line = [] }, squareIndex) => ({
  value: squares[squareIndex],
  clickHandler: squares[squareIndex]
    ? () => undefined
    : handleClick(squareIndex),
  isHighlighted: makeArrayIncludesValue(line)(squareIndex),
});
