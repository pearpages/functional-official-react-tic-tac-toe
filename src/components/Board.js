import React from 'react'

import { Square } from './Square'

export function Board({config: {squares, handleClick}}) {

  const renderSquareWithValueAndClickHandler = (square) => Square.renderSquare(squares[square], handleClick(square))

  return (
    <div>
      <div className="board-row">
        {renderSquareWithValueAndClickHandler(0)}
        {renderSquareWithValueAndClickHandler(1)}
        {renderSquareWithValueAndClickHandler(2)}
      </div>
      <div className="board-row">
        {renderSquareWithValueAndClickHandler(3)}
        {renderSquareWithValueAndClickHandler(4)}
        {renderSquareWithValueAndClickHandler(5)}
      </div>
      <div className="board-row">
        {renderSquareWithValueAndClickHandler(6)}
        {renderSquareWithValueAndClickHandler(7)}
        {renderSquareWithValueAndClickHandler(8)}
      </div>
    </div>
  );
}
