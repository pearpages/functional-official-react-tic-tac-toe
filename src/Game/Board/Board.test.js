import { render } from '@testing-library/react'

import { Board } from './Board'

describe("<Board>", () => {
  const config = {
    handleClick: () => undefined,
    line: [2, 5, 8],
    squares: [null, null, 'X', null, 'O', 'X', null, 'O', 'X']
  }

  it('should have three rows', () => {
    const result = render(<Board config={config} />)
    expect(result.container.getElementsByClassName(Board.rowClassName).length).toBe(3)
  })

  it('should call the renderer function for each cell (square), 9 times in total', () => {
    const rowRender = jest.fn()
    const boardRenderer = jest.fn(() => rowRender)
    render(<Board config={config} boardRenderer={boardRenderer} />)
    expect(rowRender).toHaveBeenCalledTimes(9)
    expect(boardRenderer).toHaveBeenCalledWith(config)
  })

  it('should warn you of any html change', () => {
    const result = render(<Board config={config} />)
    expect(result.container.innerHTML).toMatchSnapshot()
  })
})