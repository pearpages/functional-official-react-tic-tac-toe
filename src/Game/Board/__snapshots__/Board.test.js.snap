// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board> should warn you of any html change 1`] = `"<div data-testid=\\"Board\\"><div class=\\"board-row\\"><button class=\\"square \\"></button><button class=\\"square \\"></button><button class=\\"square square--highlighted\\">X</button></div><div class=\\"board-row\\"><button class=\\"square \\"></button><button class=\\"square \\">O</button><button class=\\"square square--highlighted\\">X</button></div><div class=\\"board-row\\"><button class=\\"square \\"></button><button class=\\"square \\">O</button><button class=\\"square square--highlighted\\">X</button></div></div>"`;
